### PR TITLE
[2/3] Support Open Telemetry (main content)

### PR DESCRIPTION
**Pull request by chung-shopline**
This PR contains the main content for supporting Open Telemetry:

 - Added DistributedTracingInfo field in incoming event for tracing (models/distributed_tracing_info.go)
 - Added OpenTelemetry tracing spans in multiple places
   - Dispatcher
   - Sift ~what's this?~
   - senders

Some embedded interfaces are moved away from the struct, for example:

```Go
// before
type IncomingEvent struct {
  interfaces.IncomingEventInterface
  TraceId string
  ...
}

// after
var _ interfaces.IncomingEventInterface = IncomingEvent{} // for ensuring the type implemented the interface

type IncomingEvent struct {
  TraceId string
  ...
}
```

That's because the original style is somewhat "toxic": it "helps" the `IncomingEvent` struct to implement the interface, but when the embedded member is `nil`, it actually panic at runtime when the interface method is called.

What we want is to ensure the struct implements the interface, and fail the compilation if it's not the case.

Main PR: https://github.com/shoplineapp/captin/pull/81

### CODE REVIEW

File: internal/outgoing/dispatcher.go [hunk 1]
```diff
@@ -175,6 +183,7 @@ func (d *Dispatcher) injectThrottledPayloads(ctx context.Context, e models.Incom
 		store.Remove(ctx, queueKey)
 		for _, payloadStr := range payloadStrings {
 			payload := map[string]interface{}{}
+			// TODO: handle error
```
Comment (chung-shopline) [inline internal/outgoing/dispatcher.go]: log error here

File: internal/outgoing/dispatcher.go [hunk 2]
```diff
@@ -190,6 +199,7 @@ func (d *Dispatcher) injectThrottledDocuments(ctx context.Context, e models.Inco
 		store.Remove(ctx, queueKey)
 		for _, documentStr := range documentStrings {
 			document := map[string]interface{}{}
+			// TODO: handle error
```
Comment (chung-shopline) [inline internal/outgoing/dispatcher.go]: log

File: internal/outgoing/dispatcher.go [hunk 3]
```diff
@@ -321,16 +339,20 @@ func (d *Dispatcher) processDelayedEvent(ctx context.Context, e models.IncomingE
 
 		// Schedule send event later
 		dispatcher.TrackAfterFuncJob(timeRemain, func() {
```
Comment (chung-shopline) [inline internal/outgoing/dispatcher.go:393]: maybe should make a new span for this AfterFunc, so we can simple `defer span.End()` for the original span

File: models/incoming_event.go [hunk 4]
```diff
@@ -12,15 +12,21 @@ import (
 	interfaces "github.com/shoplineapp/captin/v2/interfaces"
 )
 
+var _ interfaces.IncomingEventInterface = IncomingEvent{}
```
Comment (chung-shopline) [inline models/incoming_event.go:15]: mark the reason of moving the interface out

**Review by neal-sl — APPROVED** [review]

**Review by thomaslishopline — APPROVED** [review]

**Pull request merged**